harness/determined/pytorch/_pytorch_context.py

```diff
@@ -24,7 +25,8 @@
     import torch.cuda.amp as amp
 
     amp_import_error = False
-except ImportError:
+except ImportError:  # pragma: no cover
+    amp = None
```

Comment:
what is the purpose of `amp = None` here?

Reply:
It's to fix a complaint from PyCharm that does seem reasonable. Without the line, you don't know if a variable `amp` will be in this scope at runtime or not, a possibility made relevant by the try-except block.

Reply:
If you feel the need to fix that error, that's ok, but please fix it the right way: make `amp` into an `Optional[module]` (that is, either the real amp module or `None`), and change the if statements in the code to check if `amp` is None, instead of checking if amp_import_error is False.

Reply:
I agree with what you're asking for, but unfortunately it doesn't seem possible right now. [Here's](https://adamj.eu/tech/2021/12/29/python-type-hints-optional-imports/) a good summary of the situation.
It's much more important that we make MyPy happy than we do PyCharm, so I will revert to the flag solution here and in similar places.

Reply:
What about this:

```python
amp = None
try:
    import torch.cuda.amp

    amp = torch.cuda.amp
except ImportError:
    if torch.cuda.is_available():
        logging.warning("PyTorch AMP is unavailable.")
    pass
```

This seems to pass for me locally.  Not totally sure if it breaks mypy in other ways though.

Reply:
(also I don't have pycharm to test with pycharm)

Reply:
That code might get us around the rule that we can't use a predeclared variable in an `import` statement, but it still won't satisfy PyCharm because `torch.cuda.amp` isn't defined in the `except` block.

Reply:
got it